src/librustdoc/html/static/js/search.js

```diff
@@ -1382,16 +1425,324 @@ function initSearch(rawSearchIndex) {
+                            highlighted: prevHighlighted && value.highlighted,
+                        }, result),
+                    );
+                    return true;
+                } else if (fnType.id === typeNameIdOfFn) {
+                    writeHof(fnType, result);
+                    return true;
+                }
+                return false;
+            }
+            /**
+             * Write a type. This function checks for special types,
+             * like slices, with their own formatting. It also handles
+             * updating the where clause and generic type param map.
+             *
+             * @param {FunctionType} fnType
+             * @param {[string]} result
+             */
+            function writeFn(fnType, result) {
+                if (fnType.id < 0) {
+                    const queryId =  mgens && mgens.has(fnType.id) ? mgens.get(fnType.id) : null;
+                    if (fnParamNames[-1 - fnType.id] === "") {
+                        for (const nested of fnType.generics) {
+                            writeFn(nested, result);
+                        }
+                        return;
+                    } else if (queryId < 0) {
+                        mappedNames.set(
+                            fnParamNames[-1 - fnType.id],
+                            queryParamNames[-1 - queryId],
+                        );
+                    }
+                    pushText({
+                        name: fnParamNames[-1 - fnType.id],
+                        highlighted: !!fnType.highlighted,
+                    }, result);
+                    const where = [];
+                    onEachBtwn(
+                        fnType.generics,
+                        nested => writeFn(nested, where),
+                        () => pushText({ name: " + ", highlighted: false }, where),
+                    );
+                    if (where.length > 0) {
+                        whereClause.set(fnParamNames[-1 - fnType.id], where);
+                    }
+                } else {
+                    if (fnType.ty === TY_PRIMITIVE) {
+                        if (writeSpecialPrimitive(fnType, result)) {
+                            return;
+                        }
+                    } else if (fnType.ty === TY_TRAIT &&
+                        (fnType.id === typeNameIdOfFn ||
+                            fnType.id === typeNameIdOfFnMut ||
+                            fnType.id === typeNameIdOfFnOnce)) {
+                        writeHof(fnType, result);
+                        return;
+                    }
+                    pushText(fnType, result);
+                    if (fnType.bindings.size > 0 || fnType.generics.length > 0) {
+                        pushText({ name: "<", highlighted: false }, result);
+                    }
+                    if (fnType.bindings.size > 0) {
+                        onEachBtwn(
+                            fnType.bindings,
```

Comment:
I think it's fine to exclude `bindings` from `onEachBtwn` so the function can be a bit "better" (not having the `if` condition in the loop)

Reply:
Fine by me. Btw, did you bench with the JS changes?

Reply:
There seems to have a noticeable diff:

```
new:

Testing T -> U ... in_args = 0, returned = 0, others = 200
wall time = 1179
Testing T, U ... in_args = 0, returned = 0, others = 200
wall time = 429
Testing T -> T ... in_args = 0, returned = 0, others = 200
wall time = 555
Testing crc32 ... in_args = 0, returned = 0, others = 0
wall time = 655
Testing spi::pac ... in_args = 0, returned = 0, others = 0
wall time = 711

old:

Testing T -> U ... in_args = 0, returned = 0, others = 200
wall time = 770
Testing T, U ... in_args = 0, returned = 0, others = 200
wall time = 419
Testing T -> T ... in_args = 0, returned = 0, others = 200
wall time = 433
Testing crc32 ... in_args = 0, returned = 0, others = 0
wall time = 632
Testing spi::pac ... in_args = 0, returned = 0, others = 0
wall time = 729
```

Reply:
Do you think it's worth trying to optimize it to reduce the gap on big queries or as a follow-up PR?

Reply:
Not yet! Sorry, until https://github.com/rust-lang/rust/pull/126245 is ready, I won't have much time to take a look. Hopefully I should be "freed" from in a few hours. Then I'll gladly give a try at optimizing it away. I'm sure we can reduce the gap a little bit. (Your work on this feature remains amazing, can't wait for it to be merged!)

Reply:
Ok I'm now "free". I'll take a look tomorrow or monday. :D

Reply:
> Have you managed to run the profiler suite? It's at https://gitlab.com/notriddle/rustdoc-js-profile

After modifying the `NEW` env variable to target my local build, I can run everything, except I get:

```
ReferenceError: atob is not defined
    at new RoaringBitmap (/home/imperio/rust/rustdoc-js-profile/tmp/arti/toolchain_new/doc/static.files/search-78d69af6d344b93d.js:5:13812)
    at buildIndex (/home/imperio/rust/rustdoc-js-profile/tmp/arti/toolchain_new/doc/static.files/search-78d69af6d344b93d.js:5:16653)
    at Object.initSearch (/home/imperio/rust/rustdoc-js-profile/tmp/arti/toolchain_new/doc/static.files/search-78d69af6d344b93d.js:5:22650)
    at loadSearchJS (/home/imperio/rust/rustdoc-js-profile/src/tester.js:225:38)
    at main (/home/imperio/rust/rustdoc-js-profile/src/tester.js:311:28)
    at Object.<anonymous> (/home/imperio/rust/rustdoc-js-profile/src/tester.js:331:1)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
```

Apparently the function was removed from nodejs so needs to implement it ourselves as follows:

```js
function atob(str) {
    return Buffer.from(encodedText, 'base64').toString('utf8');
}
```

Reply:
There doesn't seem to be any error but the flamegraph is almost empty and the other files are empty. I suppose it failed somewhere but didn't provide the information... I'll continue my profiling using firefox directly.

Reply:
This is my diff btw:

```diff
diff --git a/src/main.rs b/src/main.rs
index 5f938b3..c09ac55 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -163,7 +163,7 @@ fn main() {
             .output()
             .expect("run cargo doc on new toolchain");
         if !out.status.success() {
-            panic!("stdout = {stdout:?}, stderr = {stderr:?}", stdout = String::from_utf8_lossy(&out.stdout), stderr = String::from_utf8_lossy(&out.stderr));
+            panic!("stdout = {stdout}\n\n\nstderr = {stderr}", stdout = String::from_utf8_lossy(&out.stdout), stderr = String::from_utf8_lossy(&out.stderr));
         }
         // Run old under CPU profiler 0x
         let mut cmd = Command::new("../../node_modules/.bin/0x");
diff --git a/src/tester.js b/src/tester.js
index bf8d550..7aa195d 100644
--- a/src/tester.js
+++ b/src/tester.js
@@ -4,7 +4,10 @@ const path = require("path");
 function loadContent(content) {
     const Module = module.constructor;
     const m = new Module();
-    m._compile(content, "tmp.js");
+    const extra = "function atob(str) {
+    return Buffer.from(encodedText, 'base64').toString('utf8');
+}\n";
+    m._compile(extra + content, "tmp.js");
     m.exports.ignore_order = content.indexOf("\n// ignore-order\n") !== -1 ||
         content.startsWith("// ignore-order\n");
     m.exports.exact_check = content.indexOf("\n// exact-check\n") !== -1 ||
```